File: src/main/java/com/wesker926/leetcode/algorithms/p0559/s2/Solution.java
package com.wesker926.leetcode.algorithms.p0559.s2;

import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

/**
 * @author wesker.gh
 * @date 2021/11/21
 * @description bfs
 */
public class Solution {
    public int maxDepth(Node root) {
        if (root == null) {
            return 0;
        }

        Queue<Node> queue = new LinkedList<>() {{
            add(root);
        }};
        int max = 0, count = queue.size();
        Node cur;
        while (count > 0) {
            cur = queue.poll();
            for (Node child : cur.children) {
                queue.offer(child);
            }
            count--;
            if (count == 0) {
                count = queue.size();
                max++;
            }
        }
        return max;
    }

    private static class Node {
        public int val;
        public List<Node> children;

        public Node() {
        }

        public Node(int _val) {
            val = _val;
        }

        public Node(int _val, List<Node> _children) {
            val = _val;
            children = _children;
        }
    }
}

File: src/main/java/com/wesker926/leetcode/algorithms/p0559/s2/solution.go
// Package s2
// @author: wesker.gh
// @date: 2021/11/21
// @description: bfs
package s2

type Node struct {
	Val      int
	Children []*Node
}

func maxDepth(root *Node) int {
	if root == nil {
		return 0
	}

	max, idx, count, queue := 0, 0, 1, append(make([]*Node, 0, 10000), root)
	for ; idx < len(queue); idx++ {
		for _, child := range queue[idx].Children {
			queue = append(queue, child)
		}
		count--
		if count == 0 {
			count = len(queue) - idx - 1
			max++
		}
	}
	return max
}
